File: go/500.600/514.FreedomTrail.go
package main

import (
	"fmt"
	"math"
)

func findRotateSteps(ring string, key string) int {
	n, m := len(ring), len(key)
	pos := make([][]int, 26)
	for i, c := range ring {
		pos[int(c-'a')] = append(pos[int(c-'a')], i)
	}
	dp := make([][]int, m)
	for i := range dp {
		dp[i] = make([]int, n)
		for j := range dp[i] {
			dp[i][j] = math.MaxInt
		}
	}
	for _, p := range pos[key[0]-'a'] {
		dp[0][p] = min(p, n-p) + 1
	}
	for i := 1; i < m; i++ {
		for _, j := range pos[key[i]-'a'] {
			for _, k := range pos[key[i-1]-'a'] {
				dp[i][j] = min(dp[i][j], dp[i-1][k]+min(abs(j-k), n-abs(j-k))+1)
			}
		}
	}
	return min(dp[m-1]...)
}

func min(a ...int) int {
	ret := a[0]
	for _, v := range a[1:] {
		if v < ret {
			ret = v
		}
	}
	return ret
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func main() {
	fmt.Println(findRotateSteps("caotmcaataijjxi", "oatjiioicitatajtijciocjcaaxaaatmctxamacaamjjx"), 137)
	fmt.Println(findRotateSteps("iotfo", "fioot"), 11)
	fmt.Println(findRotateSteps("abcde", "ade"), 6)
	fmt.Println(findRotateSteps("godding", "gd"), 4)
	fmt.Println(findRotateSteps("godding", "godding"), 13)
}

File: 500.600/514.FreedomTrail.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <array>
#include <unordered_map>
using namespace std;

class FreedomTrail
{
public:
    int findRotateSteps(string ring, string key)
    {
        int n = ring.size(), m = key.size();
        array<vector<int>, 26> pos;
        for (int i = 0; i < n; ++i)
        {
            pos[ring[i] - 'a'].push_back(i);
        }
        vector<vector<int>> dp(m, vector<int>(n, INT_MAX));
        for (auto &i : pos[key[0] - 'a'])
        {
            dp[0][i] = min(i, n - i) + 1;
        }
        for (int i = 1; i < m; ++i)
        {
            for (auto j : pos[key[i] - 'a'])
            {
                for (auto k : pos[key[i - 1] - 'a'])
                {
                    dp[i][j] = min(dp[i][j], dp[i - 1][k] + min(abs(j - k), n - abs(j - k)) + 1);
                }
            }
        }
        return *min_element(dp[m - 1].begin(), dp[m - 1].end());
    }
};

class FreedomTrail1
{
public:
    // 用递归求解 （会超时）
    int findRotateSteps(string ring, string key)
    {
        pos.clear();
        mRing = ring;
        mKey = key;
        n = ring.size();
        m = key.size();
        for (int i = 0; i < n; ++i)
        {
            pos[ring[i]].push_back(i);
        }
        return doring(0, 0, m);
    }

private:
    int doring(int x, int i, int ans)
    {
        if (i >= m)
        {
            return ans;
        }
        auto &p = pos[mKey[i]];
        int size = p.size();
        int d = abs(p[0] - x);
        int ret = doring(p[0], i + 1, ans + min(d, n - d));
        for (int j = 1; j < size; ++j)
        {
            int d1 = abs(p[j] - x);
            ret = min(ret, doring(p[j], i + 1, ans + min(d1, n - d1)));
        }
        return ret;
    }

private:
    unordered_map<char, vector<int>> pos;
    string mRing{};
    string mKey{};
    int n{};
    int m{};
};

int main(int argc, char const *argv[])
{
    FreedomTrail1 cls;
    cout << cls.findRotateSteps("caotmcaataijjxi", "oatjiioicitatajtijciocjcaaxaaatmctxamacaamjjx") << " -> 137" << endl;
    cout << cls.findRotateSteps("iotfo", "fioot") << " -> 11" << endl;
    cout << cls.findRotateSteps("abcde", "ade") << " -> 6" << endl;
    cout << cls.findRotateSteps("godding", "gd") << " -> 4" << endl;
    cout << cls.findRotateSteps("godding", "godding") << " -> 13" << endl;
    return 0;
}
